ibis/expr/types/relations.py

```diff
@@ -3609,6 +3609,189 @@ def pivot_wider(
+        │ a      │ a      │ a      │     1 │     1 │     1 │
+        └────────┴────────┴────────┴───────┴───────┴───────┘
+        >>> t.relocate(s.numeric(), after=s.last())
+        ┏━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━┓
+        ┃ d      ┃ e      ┃ f      ┃ a     ┃ b     ┃ c     ┃
+        ┡━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━┩
+        │ string │ string │ string │ int64 │ int64 │ int64 │
+        ├────────┼────────┼────────┼───────┼───────┼───────┤
+        │ a      │ a      │ a      │     1 │     1 │     1 │
+        └────────┴────────┴────────┴───────┴───────┴───────┘
+        >>> t.relocate(s.any_of(s.c(*"ae")))
+        ┏━━━━━━━┳━━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━━┳━━━━━━━━┓
+        ┃ a     ┃ e      ┃ b     ┃ c     ┃ d      ┃ f      ┃
+        ┡━━━━━━━╇━━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━━╇━━━━━━━━┩
+        │ int64 │ string │ int64 │ int64 │ string │ string │
+        ├───────┼────────┼───────┼───────┼────────┼────────┤
+        │     1 │ a      │     1 │     1 │ a      │ a      │
+        └───────┴────────┴───────┴───────┴────────┴────────┘
+
+        When multiple columns are selected with `before` or `after`, those
+        selected columns are moved before and after the `selectors` input
+
+        >>> t = ibis.memtable(dict(a=[1], b=["a"], c=[1], d=["a"]))
+        >>> t.relocate(s.numeric(), after=s.of_type("string"))
+        ┏━━━━━━━━┳━━━━━━━━┳━━━━━━━┳━━━━━━━┓
+        ┃ b      ┃ d      ┃ a     ┃ c     ┃
+        ┡━━━━━━━━╇━━━━━━━━╇━━━━━━━╇━━━━━━━┩
+        │ string │ string │ int64 │ int64 │
+        ├────────┼────────┼───────┼───────┤
+        │ a      │ a      │     1 │     1 │
+        └────────┴────────┴───────┴───────┘
+        >>> t.relocate(s.numeric(), before=s.of_type("string"))
+        ┏━━━━━━━┳━━━━━━━┳━━━━━━━━┳━━━━━━━━┓
+        ┃ a     ┃ c     ┃ b      ┃ d      ┃
+        ┡━━━━━━━╇━━━━━━━╇━━━━━━━━╇━━━━━━━━┩
+        │ int64 │ int64 │ string │ string │
+        ├───────┼───────┼────────┼────────┤
+        │     1 │     1 │ a      │ a      │
+        └───────┴───────┴────────┴────────┘
+        """
+        import ibis.selectors as s
+
+        if not columns and before is None and after is None and not kwargs:
+            raise com.IbisInputError(
+                "At least one selector or `before` or `after` must be provided"
+            )
+
+        if before is not None and after is not None:
+            raise com.IbisInputError("Cannot specify both `before` and `after`")
+
+        sels = {}
+        table_columns = self.columns
+
+        for name, sel in itertools.chain(
+            zip(itertools.repeat(None), map(s._to_selector, columns)),
+            zip(kwargs.keys(), map(s._to_selector, kwargs.values())),
+        ):
+            for pos in sel.positions(self):
+                if pos in sels:
+                    # make sure the last duplicate column wins by reinserting
+                    # the position if it already exists
+                    del sels[pos]
```

Comment:
I could definitely be missing something, but isn't this redundant? `sels[pos]` is going to be overwritten on the next line.

Reply:
I *think* this is to reorder the selections. The dict ordering behavior in python is that updates don't reorder, but a del and reinsert will.

```python
In [3]: d = {"a": 1, "b": 2, "c": 3}

In [4]: list(d)
Out[4]: ['a', 'b', 'c']

In [5]: d["b"] = 4

In [6]: list(d)  # order is still the same
Out[6]: ['a', 'b', 'c']

In [7]: del d["b"]

In [8]: d["b"] = 4

In [9]: list(d)  # b is now last
Out[9]: ['a', 'c', 'b']
```

Reply:
ahh, gotcha.  thanks for the explainer!  I missed that it was using the key-ordering to map the column-ordering, but that makes sense.